### PR TITLE
spec updates

### PR DESCRIPTION
**Pull request by arnetheduck**
* random small updates
* ssz no longer sorts by field, fix enum serialization
* rewire block processing a little to avoid a few state copies
* add a state simulation tool that writes out jsons

### CODE REVIEW

File: research/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+*.json
```
Comment (mratsim) [inline research/.gitignore]: I'd prefer if we settled on `./build` for all temporary outputs as we might want a default`.json` config in the repo.

Also this deals with binaries with no extension on Mac and Linux

Comment (arnetheduck) [reply research/.gitignore]: this is not a build output, it's when you run the simulator - it just dumps stuff in the same folder.. so when you do `nim c -r state_sim` while testing, it's pretty convenient

Comment (arnetheduck) [reply research/.gitignore]: and for the record, I'd prefer that too, but the infrastructure is not quite there - ie I'm hoping that output directory compiler option will land some day, then we can go full out on this

Comment (mratsim) [reply research/.gitignore]: I have a shortcut for `nim c -r -o:build/foo myproject/foo.nim` ;)

Comment (arnetheduck) [reply research/.gitignore]: yah, but it has to be something universal that even someone that just checked out the project can access

**Pull request merged**